consul/src/main/java/com/linecorp/armeria/internal/consul/ConsulClientBuilder.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.internal.consul;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nullable;
+
+import com.google.common.base.Strings;
+
+import com.linecorp.armeria.common.SessionProtocol;
+
+public class ConsulClientBuilder {
+    private static final SessionProtocol DEFAULT_CONSUL_PROTOCOL = SessionProtocol.HTTP;
+    private static final String DEFAULT_CONSUL_ADDRESS = "127.0.0.1";
+    private static final int DEFAULT_CONSUL_PORT = 8500;
+    private static final String DEFAULT_CONSUL_API_VERSION = "v1";
+    private static final Pattern CONSUL_API_VERSION_PATTERN = Pattern.compile("^v[0-9][-._a-zA-Z0-9]*$");
+
+    @Nullable
+    private URI consulUri;
+    private SessionProtocol consulProtocol = DEFAULT_CONSUL_PROTOCOL;
+    private String consulAddress = DEFAULT_CONSUL_ADDRESS;
+    private int consulPort = DEFAULT_CONSUL_PORT;
+    private String consulApiVersion = DEFAULT_CONSUL_API_VERSION;
+    @Nullable
+    private String consulToken;
+    private boolean isUriComponentSet;
+
+    protected ConsulClientBuilder() {
+    }
+
+    /**
+     * Sets the specified Consul's API service URI.
+     * The URI should include the Consul API version at path, like: {@code /v1}
+     * @param consulUri the URI of Consul API service, default: {@code http://127.0.0.1:8500/v1}
+     */
+    public ConsulClientBuilder consulUri(URI consulUri) {
```

Comment:
How about passing (`URI`)or (`Strig uri`) when creating this class and removing
`consulUri(URI consulUri)`,
`consulUri(String consulUri)`,
`consulProtocol`,
`consulAddress`,
`consulPort`,
`consulAddress`?
If we do that, we can remove the complex setting logic and simplify the uri setting.

Reply:
We have discussed this [above](https://github.com/line/armeria/pull/3002#discussion_r501575166). I think that your opinion also is not bad. Shall I change it? As you said, the version seems to have to be separated.

Reply:
Ah, I missed that. 😄 
I think if we just use the constructor which takes `Strig consulUri` that will greatly simplify the builder API. And also it will be consistent with the Eureka API.
So I prefer to use `String consulUri` but not super strong on it. So it's up to you. 😄

Reply:
Done.